Prevent chain functions from matching stored value with a rest parameter

gwhitney opened the pull request:
As discussed in #2476.

Resolves #2485.

josdejong commented on src/type/chain/Chain.js:
```diff
@@ -130,11 +130,21 @@ export const createChainClass = /* #__PURE__ */ factory(name, dependencies, ({ o
-
+      if (typed.isTypedFunction(fn)) {
+        const sigObject = typed.resolve(fn, args)
+        if (sigObject.params.length === 1) {
```
Is there no need to explicitly check if  `sigObject` has a rest parameter? 

Does this work for example for a function that has a rest parameter as second argument (like `"Matrix, ...number" or so)`? We don't have that right now but it would be neat to write this check as specific as possible.

gwhitney replied:
Right now typed-function only supports rest parameters in final position. So the only way a rest parameter could be broken across an actual first argument in the Chain and actual additional arguments in the call is if it is the only parameter. If there were more parameters, they would precede the rest parameter and the first of them would be matching the actual first argument in the Chain, meaning the rest parameter would be matching only arguments in the current call, which is OK. Thus, the only problematic case is if the signature being matched has only a single parameter (which must be a rest parameter, as it is matching multiple actual arguments, the one in the chain and at least one more from the current call, since we dispose of the case of no actual arguments in the current call earlier). Hence we need only check whether the matching signature has only a single parameter. I will put a comment to this effect and push a commit. I do not think there is any action item.

josdejong replied:
Thanks for the clarification 👍

josdejong reviewed (commented):
Looks good @gwhitney , thanks 👍 

One small inline question.

gwhitney commented:
OK, rebased and refreshed (there were some incompatible changes in types/index.d.ts as explained in the latest commit message in this branch). You can see this is now much better in sync with v11_on_v3a, and as I think I have addressed the review item with a comment explaining the rationale for this section of code, I think this can be merged into v11_on_v3a now -- and that should definitely happen before #2560 is merged into develop. Thanks for reviewing and merging if there's nothing else you see.

josdejong commented:
Merging now, thanks!
